fix(autofix/python): Fix range of expressions wrapped in parens

nmote opened the pull request:
Previously, the `$X` metavariable binding in the new fix_binop.py example was given the range associated with `1 + (5 + 3`. This is because the parentheses get dropped by the parser, so the location of the second plus operation is associated with only `5 + 3`. In isolation that's fine, but in cases like this where it is part of a larger expression, this can mean that one parenthesis is included in the range while the other is not.

This issue affects both text-based autofix and AST-based autofix. Both need to know which snippet to replace with the fix, and both need to know the correct location of the original text for bound metavariables.

My first attempt to address this was to create `ParenExpr` nodes in the generic AST, which are already used on a limited basis for Go. Even after adding cases to unwrap them in `m_expr`, there was still a fairly large difference in matching behavior. I assume that there are a lot of places outside of `m_expr` that match on expression nodes, and we'd have to add unwrapping logic everywhere that happens. This seemed error prone and hard to maintain. This is another case that illustrates the tension between an AST, most useful for matching and analysis, and a CST, most useful for manipulating and printing code.

Instead, I decided to write the range, including the parentheses, to `e_range` when the expression node was created. This is a bit of a departure from the way e_range has historically been used. Until now, as far as I know, it has only been a cache. Now, it serves as the canonical source of range information in this case. I think that this is a good direction to move in -- many of our location issues stem from some tokens not making it into the resulting AST, so I think that explicitly setting the range based on locations from the parser, when the node is constructed, will help. That way, we would be free to drop semantically meaningless tokens from the AST without worrying that they are needed to compute locations.

Fixes #2902

Test plan:

- Automated tests
- Manual test with the example from #2902 (https://semgrep.dev/playground/s/20WB)
  - Also manual test after putting the tree sitter parser first in Parse_target2.ml

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

aryx commented on libs/ast_generic/AST_generic_helpers.ml:
```diff
@@ -401,3 +401,12 @@ let undo_ac_matching_nf tok op : expr list -> expr option = function
+  | Ok l, Ok r -> e.e_range <- Some (l, r)
+  | Error _, _
+  | _, Error _ ->
+      ()
```
we might want to log the error (not that anybody will ever look at those logs ...)

aryx commented on src/fixing/Autofix.ml:
```diff
@@ -63,6 +63,8 @@ let transform_fix lang ast =
           mk_visitor
             {
               default_visitor with
+              kexpr =
```
What does this do?
need a comment. I don't get it.

nmote replied:
Oh sorry, meant to comment here but forgot... will do.

aryx reviewed (approved):
I like it. Yes we should gradually switch to use e_range.

nmote committed:
fix(autofix/python): Fix range of expressions wrapped in parens

Previously, the `$X` metavariable binding in the new fix_binop.py
example was given the range associated with `1 + (5 + 3`. This is
because the parentheses get dropped by the parser, so the location of
the second plus operation is associated with only `5 + 3`. In isolation
that's fine, but in cases like this where it is part of a larger
expression, this casn mean that one parenthesis is included in the range
while the other is not.

This issue affects both text-based autofix and AST-based autofix. Both
need to know which snippet to replace with the fix, and both need to
know the correct location of the original text for bound metavariables.

My first attempt to address this was to create `ParenExpr` nodes in the
generic AST, which are already used on a limited basis for Go. Even
after adding cases to unwrap them in `m_expr`, there was still a fairly
large difference in matching behavior. I assume that there are a lot of
places outside of `m_expr` that match on expression nodes, and we'd have
to add unwrapping logic everywhere that happens. This seemed error prone
and hard to maintain. This is another case that illustrates the tension
between an AST, most useful for matching and analysis, and a CST, most
useful for manipulating and printing code.

Instead, I decided to write the range, including the parentheses, to
`e_range` when the expression node was created. This is a bit of a
departure from the way e_range has historically been used. Until now, as
far as I know, it has only been a cache. Now, it serves as the canonical
source of range information in this case. I think that this is a good
direction to move in -- many of our location issues stem from some
tokens not making it into the resulting AST, so I think that explicitly
setting the range based on locations from the parser, when the node is
constructed, will help. That way, we would be free to drop semantically
meaningless tokens from the AST without worrying that they are needed to
compute locations.

Fixes #2902

Test plan:

- Automated tests
- Manual test with the example from #2902
  (https://semgrep.dev/playground/s/20WB)
  - Also manual test after putting the tree sitter parser first in
    Parse_target2.ml
